File: fastapi-energy-platform/app/utils/color_manager.py
"""
Color Management Utility for Energy Platform
Centralized color management system for consistent theming across the application
"""

import json
import os
import logging
from typing import Dict, List, Optional, Union
from pathlib import Path

logger = logging.getLogger(__name__)

class ColorManager:
    """
    Centralized color management system for the entire application
    Supports sectors, models, carriers, and custom color schemes
    """
    
    def __init__(self, app=None):
        self.app = app
        self.config_file = None
        self.colors = {}
        self.default_colors = self._get_default_colors()
        
        if app:
            self.init_app(app)
    
    def init_app(self, app):
        """Initialize with Flask app"""
        self.app = app
        static_folder = app.static_folder or 'static'
        self.config_file = os.path.join(static_folder, 'config', 'colors.json')
        
        # Ensure config directory exists
        os.makedirs(os.path.dirname(self.config_file), exist_ok=True)
        
        # Load existing colors or create default
        self.load_colors()
        
        # Register template functions
        app.jinja_env.globals['get_color'] = self.get_color
        app.jinja_env.globals['get_color_palette'] = self.get_color_palette
        app.jinja_env.globals['get_all_colors'] = self.get_all_colors
    
    def _get_default_colors(self) -> Dict:
        """Define comprehensive default color schemes"""
        return {
            "sectors": {
                "residential": "#2563EB",      # Blue
                "commercial": "#059669",       # Green
                "industrial": "#DC2626",       # Red
                "transportation": "#7C3AED",   # Purple
                "agriculture": "#16A34A",      # Dark Green
                "public": "#EA580C",           # Orange
                "mining": "#92400E",           # Brown
                "construction": "#0891B2",     # Cyan
                "services": "#BE185D",         # Pink
                "healthcare": "#0D9488",       # Teal
                "education": "#7C2D12",        # Dark Orange
                "hospitality": "#B91C1C",      # Dark Red
                "retail": "#1D4ED8",           # Dark Blue
                "finance": "#6B21A8",          # Dark Purple
                "technology": "#0F766E",       # Dark Teal
                "manufacturing": "#A16207",    # Amber
                "utilities": "#DC2626",        # Red
                "government": "#1F2937",       # Gray
                "military": "#374151",         # Dark Gray
                "other": "#6B7280"             # Medium Gray
            },
            "models": {
                "MLR": "#3B82F6",              # Blue
                "SLR": "#10B981",              # Emerald
                "WAM": "#F59E0B",              # Amber
                "TimeSeries": "#8B5CF6",       # Violet
                "ARIMA": "#EF4444",            # Red
                "Linear": "#06B6D4",           # Cyan
                "Polynomial": "#84CC16",       # Lime
                "Exponential": "#F97316",      # Orange
                "Logarithmic": "#EC4899",      # Pink
                "Power": "#14B8A6",            # Teal
                "Neural": "#A855F7",           # Purple
                "RandomForest": "#22C55E",     # Green
                "SVM": "#F43F5E",              # Rose
                "XGBoost": "#0EA5E9",          # Sky
                "LSTM": "#8B5CF6"              # Violet
            },
            "carriers": {
                "electricity": "#3B82F6",      # Blue
                "natural_gas": "#EF4444",      # Red
                "coal": "#1F2937",             # Gray
                "oil": "#92400E",              # Brown
                "biomass": "#16A34A",          # Green
                "solar": "#F59E0B",            # Amber
                "wind": "#06B6D4",             # Cyan
                "hydro": "#0891B2",            # Light Blue
                "nuclear": "#7C3AED",          # Purple
                "geothermal": "#DC2626",       # Red
                "waste": "#6B7280",            # Gray
                "hydrogen": "#8B5CF6",         # Violet
                "battery": "#10B981",          # Emerald
                "pumped_hydro": "#0D9488",     # Teal
                "compressed_air": "#84CC16",   # Lime
                "thermal": "#F97316",          # Orange
                "district_heating": "#EC4899", # Pink
                "district_cooling": "#14B8A6"  # Teal
            },
            "status": {
                "success": "#10B981",          # Emerald
                "warning": "#F59E0B",          # Amber
                "error": "#EF4444",            # Red
                "info": "#3B82F6",             # Blue
                "pending": "#6B7280",          # Gray
                "active": "#8B5CF6",           # Violet
                "inactive": "#9CA3AF",         # Light Gray
                "completed": "#059669",        # Dark Green
                "failed": "#DC2626",           # Dark Red
                "cancelled": "#F97316"         # Orange
            },
            "charts": {
                "primary": "#2563EB",          # Blue
                "secondary": "#059669",        # Green
                "tertiary": "#DC2626",         # Red
                "quaternary": "#7C3AED",       # Purple
                "quinary": "#EA580C",          # Orange
                "background": "#F8FAFC",       # Light Gray
                "grid": "#E2E8F0",             # Light Blue Gray
                "text": "#1E293B",             # Dark Gray
                "axis": "#64748B",             # Medium Gray
                "hover": "#F1F5F9"             # Very Light Gray
            },
            "gradients": {
                "primary": ["#3B82F6", "#1D4ED8"],     # Blue gradient
                "secondary": ["#10B981", "#059669"],    # Green gradient
                "tertiary": ["#EF4444", "#DC2626"],     # Red gradient
                "quaternary": ["#8B5CF6", "#7C3AED"],   # Purple gradient
                "success": ["#10B981", "#059669"],      # Green gradient
                "warning": ["#F59E0B", "#D97706"],      # Amber gradient
                "error": ["#EF4444", "#DC2626"],        # Red gradient
                "info": ["#3B82F6", "#2563EB"]          # Blue gradient
            },
            "themes": {
                "light": {
                    "background": "#FFFFFF",
                    "surface": "#F8FAFC",
                    "primary": "#2563EB",
                    "secondary": "#64748B",
                    "text": "#1E293B",
                    "border": "#E2E8F0"
                },
                "dark": {
                    "background": "#0F172A",
                    "surface": "#1E293B",
                    "primary": "#3B82F6",
                    "secondary": "#94A3B8",
                    "text": "#F1F5F9",
                    "border": "#334155"
                }
            }
        }
    
    def load_colors(self):
        """Load colors from JSON file or create default"""
        try:
            if os.path.exists(self.config_file):
                with open(self.config_file, 'r') as f:
                    saved_colors = json.load(f)
                # Merge with defaults to ensure all categories exist
                self.colors = self._merge_colors(self.default_colors, saved_colors)
                logger.info(f"Colors loaded from {self.config_file}")
            else:
                self.colors = self.default_colors.copy()
                self.save_colors()
                logger.info("Default colors created and saved")
        except Exception as e:
            logger.error(f"Error loading colors: {e}")
            self.colors = self.default_colors.copy()
    
    def save_colors(self):
        """Save current colors to JSON file"""
        try:
            with open(self.config_file, 'w') as f:
                json.dump(self.colors, f, indent=2)
            logger.info(f"Colors saved to {self.config_file}")
            return True
        except Exception as e:
            logger.error(f"Error saving colors: {e}")
            return False
    
    def _merge_colors(self, defaults: Dict, saved: Dict) -> Dict:
        """Merge saved colors with defaults to ensure completeness"""
        merged = defaults.copy()
        for category, colors in saved.items():
            if category in merged and isinstance(colors, dict):
                merged[category].update(colors)
            else:
                merged[category] = colors
        return merged
    
    def get_color(self, category: str, item: str, default: str = "#6B7280") -> str:
        """Get color for specific category and item"""
        try:
            return self.colors.get(category, {}).get(item, default)
        except Exception:
            return default
    
    def get_color_palette(self, category: str, items: List[str]) -> Dict[str, str]:
        """Get color palette for multiple items in a category"""
        palette = {}
        category_colors = self.colors.get(category, {})
        
        for item in items:
            if item in category_colors:
                palette[item] = category_colors[item]
            else:
                # Generate color if not exists
                palette[item] = self._generate_color_for_item(category, item)
                # Save the generated color
                if category not in self.colors:
                    self.colors[category] = {}
                self.colors[category][item] = palette[item]
        
        return palette
    
    def _generate_color_for_item(self, category: str, item: str) -> str:
        """Generate a consistent color for an item based on its name"""
        # Use hash of item name to generate consistent color
        import hashlib
        hash_object = hashlib.md5(f"{category}_{item}".encode())
        hex_dig = hash_object.hexdigest()
        
        # Convert to RGB values
        r = int(hex_dig[:2], 16)
        g = int(hex_dig[2:4], 16)
        b = int(hex_dig[4:6], 16)
        
        # Ensure good contrast and visibility
        # Adjust brightness to avoid too dark or too light colors
        brightness = (r * 299 + g * 587 + b * 114) / 1000
        if brightness < 100:  # Too dark
            r, g, b = min(255, r + 100), min(255, g + 100), min(255, b + 100)
        elif brightness > 200:  # Too light
            r, g, b = max(0, r - 100), max(0, g - 100), max(0, b - 100)
        
        return f"#{r:02X}{g:02X}{b:02X}"
    
    def set_color(self, category: str, item: str, color: str) -> bool:
        """Set color for specific category and item"""
        try:
            if category not in self.colors:
                self.colors[category] = {}
            self.colors[category][item] = color
            return self.save_colors()
        except Exception as e:
            logger.error(f"Error setting color: {e}")
            return False
    
    def set_colors(self, category: str, color_dict: Dict[str, str]) -> bool:
        """Set multiple colors for a category"""
        try:
            if category not in self.colors:
                self.colors[category] = {}
            self.colors[category].update(color_dict)
            return self.save_colors()
        except Exception as e:
            logger.error(f"Error setting colors: {e}")
            return False
    
    def get_all_colors(self) -> Dict:
        """Get all colors"""
        return self.colors.copy()
    
    def get_category_colors(self, category: str) -> Dict[str, str]:
        """Get all colors for a specific category"""
        return self.colors.get(category, {}).copy()
    
    def reset_to_defaults(self, category: Optional[str] = None) -> bool:
        """Reset colors to defaults"""
        try:
            if category:
                if category in self.default_colors:
                    self.colors[category] = self.default_colors[category].copy()
            else:
                self.colors = self.default_colors.copy()
            return self.save_colors()
        except Exception as e:
            logger.error(f"Error resetting colors: {e}")
            return False
    
    def get_chart_colors(self, count: int, category: str = "charts") -> List[str]:
        """Get a list of colors for charts"""
        base_colors = [
            self.get_color(category, "primary"),
            self.get_color(category, "secondary"), 
            self.get_color(category, "tertiary"),
            self.get_color(category, "quaternary"),
            self.get_color(category, "quinary")
        ]
        
        # If we need more colors, generate them
        if count > len(base_colors):
            additional_colors = []
            for i in range(len(base_colors), count):
                color = self._generate_color_for_item("chart", f"color_{i}")
                additional_colors.append(color)
            base_colors.extend(additional_colors)
        
        return base_colors[:count]
    
    def get_sector_colors(self, sectors: List[str]) -> Dict[str, str]:
        """Get colors specifically for sectors"""
        return self.get_color_palette("sectors", sectors)
    
    def get_model_colors(self, models: List[str]) -> Dict[str, str]:
        """Get colors specifically for models"""
        return self.get_color_palette("models", models)
    
    def get_carrier_colors(self, carriers: List[str]) -> Dict[str, str]:
        """Get colors specifically for carriers"""
        return self.get_color_palette("carriers", carriers)
    
    def export_colors_for_js(self) -> str:
        """Export colors as JavaScript object"""
        try:
            return f"window.AppColors = {json.dumps(self.colors, indent=2)};"
        except Exception as e:
            logger.error(f"Error exporting colors for JS: {e}")
            return "window.AppColors = {};"
    
    def get_gradient(self, gradient_name: str) -> List[str]:
        """Get gradient colors"""
        return self.colors.get("gradients", {}).get(gradient_name, ["#3B82F6", "#1D4ED8"])
    
    def get_theme_colors(self, theme: str = "light") -> Dict[str, str]:
        """Get theme colors"""
        return self.colors.get("themes", {}).get(theme, self.colors["themes"]["light"])

# Global instance
color_manager = ColorManager()

def init_color_manager(app):
    """Initialize color manager with Flask app"""
    color_manager.init_app(app)
    return color_manager

# Utility functions for direct use
def get_color(category: str, item: str, default: str = "#6B7280") -> str:
    """Direct function to get color"""
    return color_manager.get_color(category, item, default)

def get_sector_colors(sectors: List[str]) -> Dict[str, str]:
    """Direct function to get sector colors"""
    return color_manager.get_sector_colors(sectors)

def get_model_colors(models: List[str]) -> Dict[str, str]:
    """Direct function to get model colors"""
    return color_manager.get_model_colors(models)

def get_chart_colors(count: int) -> List[str]:
    """Direct function to get chart colors"""
    return color_manager.get_chart_colors(count)

File: static/js/color_manager.js
/**
 * Color Manager for Energy Platform
 * Dynamic color management with Chart.js 4.4.0 integration and real-time updates
 */

class ColorManager {
    constructor() {
        this.colors = {};
        this.initialized = false;
        this.apiBase = '/demand_visualization/api/colors';
        this.changeCallbacks = new Set();
        this.currentTheme = 'light';
        this.availableThemes = ['light', 'dark'];
        this.isUpdating = false;
        
        this.init();
    }

    async init() {
        try {
            await this.loadColors();
            await this.loadThemeInfo();
            this.setupEventListeners();
            this.initialized = true;
            console.log('✅ Color Manager initialized successfully');
        } catch (error) {
            console.error('❌ Color Manager initialization failed:', error);
            this.loadFallbackColors();
        }
    }

    async loadColors() {
        try {
            // Try to get colors from window.AppColors first (set by template)
            if (window.AppColors && typeof window.AppColors === 'object') {
                this.colors = JSON.parse(JSON.stringify(window.AppColors));
                console.log('🎨 Colors loaded from window.AppColors');
                return;
            }

            // Fallback to API
            const response = await fetch(`${this.apiBase}/get-all`);
            if (response.ok) {
                const result = await response.json();
                if (result.success && result.data) {
                    this.colors = result.data;
                    this.currentTheme = result.data.current_theme || 'light';
                    this.availableThemes = result.data.available_themes || ['light', 'dark'];
                    console.log('🎨 Colors loaded from API');
                } else {
                    throw new Error(result.error || 'API returned no data');
                }
            } else {
                throw new Error(`HTTP ${response.status}`);
            }
        } catch (error) {
            console.warn('⚠️ Color loading failed, using fallback:', error);
            this.loadFallbackColors();
        }
    }

    async loadThemeInfo() {
        try {
            const response = await fetch('/api/charts/themes');
            if (response.ok) {
                const result = await response.json();
                if (result.success) {
                    this.currentTheme = result.current_theme;
                    this.availableThemes = result.available_themes;
                }
            }
        } catch (error) {
            console.warn('Theme info loading failed:', error);
        }
    }

    loadFallbackColors() {
        this.colors = {
            sectors: {
                residential: "#2563EB",
                commercial: "#059669", 
                industrial: "#DC2626",
                transportation: "#7C3AED",
                agriculture: "#16A34A",
                public: "#EA580C",
                mining: "#92400E",
                construction: "#0891B2",
                services: "#BE185D",
                healthcare: "#0D9488",
                education: "#7C2D12",
                other: "#6B7280"
            },
            models: {
                MLR: "#3B82F6",
                SLR: "#10B981", 
                WAM: "#F59E0B",
                TimeSeries: "#8B5CF6",
                ARIMA: "#EF4444",
                Linear: "#06B6D4",
                Polynomial: "#84CC16",
                Exponential: "#F97316",
                Logarithmic: "#EC4899",
                Neural: "#A855F7",
                Custom: "#6B7280"
            },
            charts: {
                primary: "#2563EB",
                secondary: "#059669",
                tertiary: "#DC2626", 
                quaternary: "#7C3AED",
                quinary: "#EA580C"
            },
            status: {
                success: "#10B981",
                warning: "#F59E0B",
                error: "#EF4444",
                info: "#3B82F6"
            }
        };
        this.initialized = true;
        console.log('🔧 Fallback colors loaded');
    }

    setupEventListeners() {
        // Listen for theme changes
        document.addEventListener('themeChanged', (event) => {
            this.handleThemeChange(event.detail.theme);
        });

        // Listen for scenario changes to load scenario-specific colors
        document.addEventListener('scenarioChanged', (event) => {
            this.loadScenarioColors(event.detail.scenarioName);
        });
    }

    // ========== COLOR RETRIEVAL METHODS ==========

    getColor(category, item, defaultColor = "#6B7280") {
        if (!this.initialized) {
            console.warn('⚠️ Color Manager not initialized, using default');
            return defaultColor;
        }

        try {
            const categoryColors = this.colors[category];
            if (categoryColors && categoryColors[item]) {
                return categoryColors[item];
            }

            // Generate color if not exists and add to colors
            if (categoryColors) {
                const generatedColor = this.generateColorForItem(category, item);
                categoryColors[item] = generatedColor;
                return generatedColor;
            }

            return defaultColor;
        } catch (error) {
            console.error(`❌ Error getting color for ${category}.${item}:`, error);
            return defaultColor;
        }
    }

    getColorPalette(category, items) {
        const palette = {};
        items.forEach(item => {
            palette[item] = this.getColor(category, item);
        });
        return palette;
    }

    getSectorColors(sectors) {
        if (!Array.isArray(sectors)) {
            console.warn('getSectorColors expects an array');
            return {};
        }
        return this.getColorPalette('sectors', sectors);
    }

    getModelColors(models) {
        if (!Array.isArray(models)) {
            console.warn('getModelColors expects an array');
            return {};
        }
        return this.getColorPalette('models', models);
    }

    getChartColors(count, startIndex = 0) {
        const baseColors = [
            this.getColor('charts', 'primary'),
            this.getColor('charts', 'secondary'),
            this.getColor('charts', 'tertiary'),
            this.getColor('charts', 'quaternary'), 
            this.getColor('charts', 'quinary')
        ];

        const colors = [];
        for (let i = 0; i < count; i++) {
            const colorIndex = (startIndex + i) % baseColors.length;
            if (colorIndex < baseColors.length) {
                colors.push(baseColors[colorIndex]);
            } else {
                // Generate additional colors
                colors.push(this.generateColorForItem('chart', `generated_${i}`));
            }
        }

        return colors;
    }

    getStatusColor(status) {
        return this.getColor('status', status);
    }

    getThemeColor(colorKey) {
        return this.getColor('themes', `${this.currentTheme}_${colorKey}`);
    }

    // ========== COLOR GENERATION ==========

    generateColorForItem(category, item) {
        // color generation with better distribution
        const str = `${category}_${item}`;
        let hash = 0;
        
        for (let i = 0; i < str.length; i++) {
            hash = ((hash << 5) - hash + str.charCodeAt(i)) & 0xffffffff;
        }

        hash = Math.abs(hash);

        // Use HSL for better color control
        const hue = hash % 360;
        const saturation = 60 + (hash % 30); // 60-90%
        const lightness = 45 + (hash % 25);  // 45-70%

        return this.hslToHex(hue, saturation, lightness);
    }

    hslToHex(h, s, l) {
        h = h / 360;
        s = s / 100; 
        l = l / 100;

        const hue2rgb = (p, q, t) => {
            if (t < 0) t += 1;
            if (t > 1) t -= 1;
            if (t < 1/6) return p + (q - p) * 6 * t;
            if (t < 1/2) return q;
            if (t < 2/3) return p + (q - p) * (2/3 - t) * 6;
            return p;
        };

        let r, g, b;

        if (s === 0) {
            r = g = b = l; // achromatic
        } else {
            const q = l < 0.5 ? l * (1 + s) : l + s - l * s;
            const p = 2 * l - q;
            r = hue2rgb(p, q, h + 1/3);
            g = hue2rgb(p, q, h);
            b = hue2rgb(p, q, h - 1/3);
        }

        const toHex = (c) => {
            const hex = Math.round(c * 255).toString(16);
            return hex.length === 1 ? '0' + hex : hex;
        };

        return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
    }

    // ========== COLOR MODIFICATION ==========

    async updateColor(category, item, color) {
        if (this.isUpdating) {
            console.warn('Color update already in progress');
            return false;
        }

        this.isUpdating = true;

        try {
            // Validate color format
            if (!this.isValidHexColor(color)) {
                throw new Error('Invalid color format');
            }

            // Update locally first
            if (!this.colors[category]) {
                this.colors[category] = {};
            }
            
            const oldColor = this.colors[category][item];
            this.colors[category][item] = color;

            // Try to save to backend
            try {
                await this.saveColorToBackend(category, item, color);
                console.log(`✅ Color updated: ${category}.${item} = ${color}`);
            } catch (error) {
                console.warn('⚠️ Backend save failed, keeping local change:', error);
            }

            // Notify all callbacks
            this.notifyColorChange(category, item, color, oldColor);

            return true;
        } catch (error) {
            console.error('❌ Color update failed:', error);
            return false;
        } finally {
            this.isUpdating = false;
        }
    }

    async updateMultipleColors(category, colorDict) {
        const updates = [];
        
        for (const [item, color] of Object.entries(colorDict)) {
            updates.push(this.updateColor(category, item, color));
        }

        const results = await Promise.all(updates);
        return results.every(result => result === true);
    }

    async saveColorToBackend(category, item, color) {
        const response = await fetch(`${this.apiBase}/set`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ category, item, color })
        });

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.error || `HTTP ${response.status}`);
        }

        const result = await response.json();
        if (!result.success) {
            throw new Error(result.error || 'Backend reported failure');
        }
    }

    isValidHexColor(color) {
        return /^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/.test(color);
    }

    // ========== SCENARIO-SPECIFIC COLORS ==========

    async loadScenarioColors(scenarioName) {
        if (!scenarioName) return;

        try {
            const response = await fetch(`${this.apiBase}/scenario/${scenarioName}`);
            if (response.ok) {
                const result = await response.json();
                if (result.success && result.data) {
                    // Merge scenario-specific colors
                    if (result.data.sectors) {
                        this.colors.sectors = { ...this.colors.sectors, ...result.data.sectors };
                    }
                    if (result.data.models) {
                        this.colors.models = { ...this.colors.models, ...result.data.models };
                    }
                    
                    console.log(`🎨 Loaded colors for scenario: ${scenarioName}`);
                    this.notifyColorsReloaded(scenarioName);
                }
            }
        } catch (error) {
            console.warn(`Failed to load scenario colors for ${scenarioName}:`, error);
        }
    }

    // ========== THEME MANAGEMENT ==========

    async setTheme(themeName) {
        if (!this.availableThemes.includes(themeName)) {
            console.error(`Theme '${themeName}' not available`);
            return false;
        }

        try {
            const response = await fetch('/chart_management/themes/custom', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ theme: themeName })
            });

            if (response.ok) {
                const result = await response.json();
                if (result.success) {
                    this.currentTheme = themeName;
                    await this.loadColors(); // Reload colors for new theme
                    this.notifyThemeChange(themeName);
                    return true;
                }
            }
            return false;
        } catch (error) {
            console.error('Theme change failed:', error);
            return false;
        }
    }

    getCurrentTheme() {
        return this.currentTheme;
    }

    getAvailableThemes() {
        return [...this.availableThemes];
    }

    // ========== COLOR UTILITIES ==========

    addTransparency(hexColor, alpha) {
        try {
            const hex = hexColor.replace('#', '');
            const r = parseInt(hex.substr(0, 2), 16);
            const g = parseInt(hex.substr(2, 2), 16);
            const b = parseInt(hex.substr(4, 2), 16);
            return `rgba(${r}, ${g}, ${b}, ${alpha})`;
        } catch (error) {
            console.warn('Failed to add transparency:', error);
            return `rgba(99, 102, 241, ${alpha})`;
        }
    }

    darkenColor(hexColor, factor) {
        try {
            const hex = hexColor.replace('#', '');
            let r = parseInt(hex.substr(0, 2), 16);
            let g = parseInt(hex.substr(2, 2), 16);
            let b = parseInt(hex.substr(4, 2), 16);

            r = Math.max(0, Math.floor(r * (1 - factor)));
            g = Math.max(0, Math.floor(g * (1 - factor)));
            b = Math.max(0, Math.floor(b * (1 - factor)));

            return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
        } catch (error) {
            console.warn('Failed to darken color:', error);
            return '#1D4ED8';
        }
    }

    lightenColor(hexColor, factor) {
        try {
            const hex = hexColor.replace('#', '');
            let r = parseInt(hex.substr(0, 2), 16);
            let g = parseInt(hex.substr(2, 2), 16);
            let b = parseInt(hex.substr(4, 2), 16);

            r = Math.min(255, Math.floor(r + (255 - r) * factor));
            g = Math.min(255, Math.floor(g + (255 - g) * factor));
            b = Math.min(255, Math.floor(b + (255 - b) * factor));

            return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
        } catch (error) {
            console.warn('Failed to lighten color:', error);
            return '#93C5FD';
        }
    }

    getContrastColor(hexColor) {
        try {
            const hex = hexColor.replace('#', '');
            const r = parseInt(hex.substr(0, 2), 16);
            const g = parseInt(hex.substr(2, 2), 16);
            const b = parseInt(hex.substr(4, 2), 16);

            // Calculate relative luminance
            const luminance = (0.299 * r + 0.587 * g + 0.114 * b) / 255;
            
            return luminance > 0.5 ? '#000000' : '#FFFFFF';
        } catch (error) {
            return '#000000';
        }
    }

    // ========== CHART.JS INTEGRATION ==========

    createChartDataset(label, data, options = {}) {
        const {
            category = 'charts',
            item = 'primary',
            chartType = 'line',
            fill = false,
            tension = 0.2,
            pointRadius = 4,
            borderWidth = 2
        } = options;

        const color = this.getColor(category, item);

        const dataset = {
            label: label,
            data: data,
            borderColor: color,
            backgroundColor: fill ? this.addTransparency(color, 0.3) : this.addTransparency(color, 0.1),
            fill: fill,
            tension: tension,
            borderWidth: borderWidth,
            pointRadius: pointRadius,
            pointHoverRadius: pointRadius + 2,
            pointBackgroundColor: '#FFFFFF',
            pointBorderColor: color,
            pointBorderWidth: 2
        };

        // Chart type specific styling
        if (chartType === 'bar') {
            dataset.backgroundColor = this.addTransparency(color, 0.8);
            dataset.borderRadius = 4;
            dataset.borderSkipped = false;
            // Remove point styling for bars
            delete dataset.pointRadius;
            delete dataset.pointHoverRadius;
            delete dataset.pointBackgroundColor;
            delete dataset.pointBorderColor;
            delete dataset.pointBorderWidth;
        } else if (chartType === 'area') {
            dataset.fill = true;
            dataset.backgroundColor = this.addTransparency(color, 0.3);
        }

        return dataset;
    }

    enhanceChartConfig(chartConfig, colorMapping = {}) {
        if (!chartConfig.data?.datasets) return chartConfig;

        chartConfig.data.datasets.forEach((dataset, index) => {
            const label = dataset.label || '';
            
            // Try to get appropriate color based on mapping or label
            let color;
            if (colorMapping[label]) {
                color = colorMapping[label];
            } else if (this.colors.sectors && this.colors.sectors[label.toLowerCase()]) {
                color = this.colors.sectors[label.toLowerCase()];
            } else if (this.colors.models && this.colors.models[label]) {
                color = this.colors.models[label];
            } else {
                color = this.getChartColors(1, index)[0];
            }

            // Apply colors
            dataset.borderColor = color;
            
            if (chartConfig.type === 'bar') {
                dataset.backgroundColor = this.addTransparency(color, 0.8);
                dataset.borderColor = this.darkenColor(color, 0.1);
            } else if (chartConfig.type === 'line' || chartConfig.type === 'area') {
                dataset.backgroundColor = this.addTransparency(color, 
                    chartConfig.type === 'area' ? 0.3 : 0.1);
                dataset.pointBackgroundColor = '#FFFFFF';
                dataset.pointBorderColor = color;
            }
        });

        return chartConfig;
    }

    // ========== EVENT MANAGEMENT ==========

    registerColorChangeCallback(callback) {
        if (typeof callback === 'function') {
            this.changeCallbacks.add(callback);
        }
    }

    unregisterColorChangeCallback(callback) {
        this.changeCallbacks.delete(callback);
    }

    notifyColorChange(category, item, newColor, oldColor) {
        const changeEvent = {
            category,
            item,
            newColor,
            oldColor,
            timestamp: Date.now()
        };

        this.changeCallbacks.forEach(callback => {
            try {
                callback(changeEvent);
            } catch (error) {
                console.error('Color change callback error:', error);
            }
        });

        // Dispatch custom event
        document.dispatchEvent(new CustomEvent('colorChanged', {
            detail: changeEvent
        }));
    }

    notifyColorsReloaded(scenarioName) {
        document.dispatchEvent(new CustomEvent('colorsReloaded', {
            detail: { scenarioName, timestamp: Date.now() }
        }));
    }

    notifyThemeChange(themeName) {
        document.dispatchEvent(new CustomEvent('themeChanged', {
            detail: { theme: themeName, timestamp: Date.now() }
        }));
    }

    // ========== PERSISTENCE AND RESET ==========

    async resetToDefaults(category = null) {
        try {
            const response = await fetch(`${this.apiBase}/reset`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ category })
            });

            if (response.ok) {
                const result = await response.json();
                if (result.success) {
                    await this.loadColors(); // Reload colors
                    console.log(`✅ Colors reset: ${category || 'all'}`);
                    
                    // Notify about reset
                    document.dispatchEvent(new CustomEvent('colorsReset', {
                        detail: { category, timestamp: Date.now() }
                    }));
                    
                    return true;
                }
            }
            return false;
        } catch (error) {
            console.error('Reset failed:', error);
            return false;
        }
    }

    async saveAllColors() {
        try {
            const response = await fetch(`${this.apiBase}/save-all`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(this.colors)
            });

            if (response.ok) {
                const result = await response.json();
                if (result.success) {
                    console.log('✅ All colors saved successfully');
                    return true;
                }
            }
            return false;
        } catch (error) {
            console.error('Save all colors failed:', error);
            return false;
        }
    }

    // ========== UTILITY METHODS ==========

    getAllColors() {
        return JSON.parse(JSON.stringify(this.colors));
    }

    getCategoryColors(category) {
        return { ...(this.colors[category] || {}) };
    }

    async waitForInitialization() {
        if (this.initialized) return;

        let attempts = 0;
        const maxAttempts = 100;

        while (!this.initialized && attempts < maxAttempts) {
            await new Promise(resolve => setTimeout(resolve, 100));
            attempts++;
        }

        if (!this.initialized) {
            console.warn('⚠️ Color Manager initialization timeout');
            this.loadFallbackColors();
        }
    }

    getColorInfo() {
        return {
            initialized: this.initialized,
            currentTheme: this.currentTheme,
            availableThemes: this.availableThemes,
            totalColors: Object.keys(this.colors).reduce((sum, category) => {
                return sum + Object.keys(this.colors[category] || {}).length;
            }, 0),
            categories: Object.keys(this.colors)
        };
    }
}

// ========== GLOBAL INITIALIZATION ==========

// Create global instance
window.ColorManager = ColorManager;
window.colorManager = new ColorManager();

// Global utility functions for backward compatibility
window.getColor = (category, item, defaultColor) => {
    return window.colorManager.getColor(category, item, defaultColor);
};

window.getSectorColors = (sectors) => {
    return window.colorManager.getSectorColors(sectors);
};

window.getModelColors = (models) => {
    return window.colorManager.getModelColors(models);
};

window.getChartColors = (count, startIndex = 0) => {
    return window.colorManager.getChartColors(count, startIndex);
};

window.updateColor = async (category, item, color) => {
    return window.colorManager.updateColor(category, item, color);
};

// global functions
window.enhanceChartConfig = (chartConfig, colorMapping) => {
    return window.colorManager.enhanceChartConfig(chartConfig, colorMapping);
};

window.createChartDataset = (label, data, options) => {
    return window.colorManager.createChartDataset(label, data, options);
};

// Export for modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = ColorManager;
}

console.log('🎨 Color Manager loaded successfully');